### DISCUSSION TITLE
Issue & fix for CoApp, .json file location (Mac)

### DISCUSSION THREAD

**Discussion by paultillery** [aclap-dev/video-downloadhelper#291] [Help and Q&A]
### Your question

Hello,

I recently had an issue with VDH, and I found a fix I hadn’t seen in the troubleshooting materials (apologies if I overlooked it). Since my issue's now resolved, I hope this might be helpful to others, and I’m curious if it stemmed from user-error I can avoid repeating. Here’s my info:

**Platform:** Mac aarch64 (macOS Sonoma 14.2.1)
**Browser:** Tor Project Firefox 115.8.0

**TL;DR (best I understand):**
- **Issue:** after updating the companion app, the net.downloadhelper.coapp.json file only updated in my user Library, but needed to update in the system Library.
- **Fix:** overwriting the .json file in the system Library with a copy from my user Library resolved the issue for me.

**Details (this was Tue. 3/12/24):**
- When I tried to use VDH, it said I needed to update the companion app from 1.6.3 to 2.0.19.
- I installed VdhCoApp 2.0.19, launched it, and saw “VdhCoApp is ready to be used.”
- VDH still did not work, and still prompted me to install the companion app.
- The Extension Settings said “Checking companion app returned: *An unexpected error occurred.*”
- In the system Library (/Library/Application Support/Mozilla/NativeMessagingHosts), the file net.downloadhelper.coapp.json included a "path" line that seemed outdated:

> "path": “/Applications/net.downloadhelper.coapp.app/Contents/MacOS/bin/net.downloadhelper.coapp-mac-64"
> 
> (Correct through "MacOS", but the "/bin" subfolder no longer existed)

- In my user Library (/Users/[myusername]/Library/Application Support/Mozilla/NativeMessagingHosts), there was a newer net.downloadhelper.coapp.json whose "path" line was correct:

> "path": “/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp"
> 
> (This matched the post-update contents of net.downloadhelper.coapp.app)

- I copied the user Library's newer .json and used it to overwrite the system Library's older .json.
- After that, VDH worked again!

Again, I hope this info might be helpful, and I’d welcome insights on whether user-error was a factor. And if any dev(s) see this, thanks for making a very useful extension!

### Addon & CoApp version and details

version: 8.2.1.1
channel: stable
build date: 2024-02-19
build options: linuxlic: false, noyt: false, browser: firefox
lang: en-US
license: unset
platform: aarch64 mac
UA: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/115.0
coapp: yes. Info:
{
  "id": "net.downloadhelper.coapp",
  "name": "VdhCoApp",
  "version": "2.0.19",
  "binary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/vdhcoapp",
  "displayName": "VdhCoApp",
  "description": "Video DownloadHelper companion app",
  "target": {
    "os": "mac",
    "arch": "arm64",
    "node": "18"
  },
  "home": "/Users/paultilleryiv",
  "converterBinary": "/Applications/net.downloadhelper.coapp.app/Contents/MacOS/ffmpeg",
  "converterBase": "ffmpeg",
  "converterBaseVersion": "n6.1-11-gc30f360955"
}

**Comment by GrampaWildWilly** [comment]
I don't work for VDH.  I'm just another user.

You upgraded from a CoApp version 1.x to a CoApp version 2.x.  It is true that the location of the CoApp under the 2.x versions has changed.  I think you may have uncovered an error in the installation process.  I'm on Windows & didn't encounter this issue so my guess is it's an issue particular to Mac.

**Reply by paultillery** [reply]
Thanks for the reply! Sounds about right, though I'm keeping an eye out for any settings I might be able to change (beyond turning off automatic updates) to avoid this in the future.

**Reply by GrampaWildWilly** [reply]
Turning off automatic updates is something you should do on general principles.  A few users have reported that VDH got updated unexpectedly while VDH was downloading something.  This resulted in the download being immediately, abruptly cancelled, leaving behind an unplayable file.  The only recourse was to redo the download from the beginning.  Those users were justifiably upset.  But they incorrectly blamed VDH for the problem.  The fact is that the VDH software itself does not update VDH on your system.  The VDH devs don't update VDH on your system, either.  Firefox performs the update.  Firefox performs the update according to settings.  You control settings.  So ultimately, it's the user's fault if Firefox performs an automatic update at an inopportune time.  Due to the nature of VDH, you should not allow Firefox to automatically update VDH.  This might not be the same decision you make for other add-ons, but it is definitely the decision you should make for VDH.  This is an issue that @paulrouget is looking into.  But until he provides that solution, you should turn off automatic updates for VDH, both for the general availability version & for the beta.  It does introduce the minor inconvenience that you need to remember to go check for updates yourself every so often.  But then, you can make sure such updates do NOT happen during a download you might be doing.  You can decide to install the update on your schedule, or not at all, your choice.  So turning off automatic updates for VDH is definitely what you should do now.

Meanwhile, I still believe you have uncovered an error in the installation process.  Paul would be better able to determine what to do about it.  My feeling is there's no settings within VDH that could affect this.  VDH settings control the behavior of the product once it's installed.  They don't have any effect on the installation process.

**Reply by paultillery** [reply]
Appreciate the insight! I've now turned off auto-updates; thanks again.

**Comment by nirav24x7** [comment]
I only encounter an issue on Tor but not on Firefox on the same computer so it maybe related to the installation as well as the Tor browser
